storage/pebble/bootstrap.go

```diff
@@ -0,0 +1,141 @@
+	}
+	return &Bootstrap{
+		checkpointDir:      checkpointDir,
+		checkpointFileName: checkpointFileName,
+		log:                log,
+		db:                 db,
+		leafNodeChan:       make(chan *wal.LeafNode, checkpointLeafNodeBufSize),
+		rootHeight:         rootHeight,
+	}, nil
+}
+
+func (b *Bootstrap) batchIndexRegisters(leafNodes []*wal.LeafNode) error {
+	batch := b.db.NewBatch()
+	defer batch.Close()
+	for _, register := range leafNodes {
+		payload := register.Payload
+		key, err := payload.Key()
+		if err != nil {
+			return fmt.Errorf("could not get key from register payload: %w", err)
+		}
+
+		registerID, err := keyToRegisterID(key)
+		if err != nil {
+			return fmt.Errorf("could not get register ID from key: %w", err)
+		}
+
+		encoded := newLookupKey(b.rootHeight, registerID).Bytes()
+		err = batch.Set(encoded, payload.Value(), nil)
+		if err != nil {
+			return fmt.Errorf("failed to set key: %w", err)
+		}
+	}
+	err := batch.Commit(pebble.Sync)
+	if err != nil {
+		return fmt.Errorf("failed to commit batch: %w", err)
+	}
+	return nil
+}
+
+// indexCheckpointFileWorker asynchronously indexes register entries in b.checkpointDir
+// with wal.OpenAndReadLeafNodesFromCheckpointV6
+func (b *Bootstrap) indexCheckpointFileWorker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+	select {
+	case <-ctx.Done():
+		return
+	default:
+	}
+	// collect leaf nodes to batch index until the channel is closed
+	batch := make([]*wal.LeafNode, 0, pebbleBootstrapRegisterBatchLen)
+	for leafNode := range b.leafNodeChan {
+		batch = append(batch, leafNode)
+		if len(batch) >= pebbleBootstrapRegisterBatchLen {
+			err := b.batchIndexRegisters(batch)
+			if err != nil {
+				ctx.Throw(fmt.Errorf("unable to index registers to pebble in batch: %w", err))
+			}
+			batch = make([]*wal.LeafNode, 0, pebbleBootstrapRegisterBatchLen)
+		}
+	}
+	err := b.batchIndexRegisters(batch)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("unable to index remaining registers to pebble: %w", err))
+	}
+}
+
+// IndexCheckpointFile indexes the checkpoint file in the Dir provided as a component
+func (b *Bootstrap) IndexCheckpointFile() error {
+	ctx := context.Background()
+	sigCtx, errChan := irrecoverable.WithSignaler(ctx)
+	// index checkpoint file async
+	cmb := component.NewComponentManagerBuilder()
+	for i := 0; i < pebbleBootstrapWorkerCount; i++ {
+		// create workers to read and index registers
+		cmb.AddWorker(b.indexCheckpointFileWorker)
+	}
+	err := wal.OpenAndReadLeafNodesFromCheckpointV6(b.leafNodeChan, b.checkpointDir, b.checkpointFileName, b.log)
+	if err != nil {
+		// error in reading a leaf node
+		return fmt.Errorf("error reading leaf node: %w", err)
+	}
+	c := cmb.Build()
+	c.Start(sigCtx)
+	// wait for the indexing to finish before populating heights
+	<-c.Done()
+
+	select {
+	case procErr := <-errChan:
+		return fmt.Errorf("failed to index checkpoint file: %w", procErr)
+	default:
+	}
```

Comment:
I'm not sure if `Done()` will ever be signaled if an error is returned. Looking at the code, the logic that closes the channel returned by `Done()` is run in a defer.
https://github.com/onflow/flow-go/blob/30ae08b71103b93edbc85b447b2f330827daa5cc/module/component/component.go#L222-L242

However, the code that throw's the error on the error channel is called first. If you look at the logic of `Throw`:
https://github.com/onflow/flow-go/blob/30ae08b71103b93edbc85b447b2f330827daa5cc/module/irrecoverable/irrecoverable.go#L31-L41

The function ends with calling `os.Exit(0)`, killing the gorountine likely before that original defer is run. You could try it out with some tests, but I suspect it is not called.

Instead, you can just wait for either done or an error here. There's also a util function that handles this along with the race conditions for you

https://github.com/onflow/flow-go/blob/30ae08b71103b93edbc85b447b2f330827daa5cc/module/util/util.go#L125-L146

```suggestion
	// wait for the indexing to finish before populating heights
	if err := util.WaitError(errChan, c.Done()); err != nil {
		return fmt.Errorf("failed to index checkpoint file: %w", err)
	}
```